Allow cleanup on closing more-info

andrey-git opened the pull request:
Remove camera's streaming source from camera's 'more info' when more-info is closed.

andrey-git committed:
Remove camera's streaming source from camera's 'more info' when more-info is closed.

balloob commented on src/more-infos/more-info-content.html:
```diff
@@ -35,11 +35,14 @@
-      { hass: this.hass, stateObj: stateObj });
+    if (!stateObj) {
+      window.hassUtil.dynamicContentUpdater(
+        this, null, { cleanup: true });
```
Can you rename it to `isVisible`, which is more in line with what it represents

andrey-git committed:
Rename to isVisible

andrey-git committed:
Rename to isVisible

balloob commented on src/more-infos/more-info-camera.html:
```diff
@@ -29,21 +29,26 @@
     if (hass.demo) {
       return '/demo/webcam.jpg';
-    } else if (stateObj) {
+    } else if (stateObj && isVisible) {
```
I'm wondering now, I thought that this logic of `isVisible` was provided by the fact that stateObj is set to null when the more info dialog is hidden, is that no longer the case?

andrey-git replied:
The call is "window.hassUtil.dynamicContentUpdater(this, null, { isVisible: false })"
And dynamicContentUpdater doesn't clear attributes which are not passed, so just isVisible is changed to false but, stateObj keeps its previous value.

At first I though just to pass stateObj:null instead of adding a new attribute, but some other classes expect stateObj to be non null.

balloob commented on src/more-infos/more-info-camera.html:
```diff
@@ -29,21 +29,26 @@
              '?token=' + stateObj.attributes.access_token;
     }
-    // Return an empty image if no stateObj (= dialog not open)
+    // Return an empty image if no stateObj (= dialog not open) or in cleanup mode.
```
This comment is no longer correct, stateObj will not be null.

andrey-git replied:
stateObj can still be null is some flow I didn't change. (i.e. not from dynamicContentUpdater)
In that case camera was using the fallback image and will still do so after my change.

balloob commented on src/more-infos/more-info-content.html:
```diff
@@ -35,11 +35,14 @@
-      { hass: this.hass, stateObj: stateObj });
+    if (!stateObj) {
+      window.hassUtil.dynamicContentUpdater(
+        this, null, { isVisible: false });
```
by setting it to null, if we close a more info light and open it again, it has to re-create all the DOM. The old logic was faster. 

Also, no need to pass attributes if you're not going to render anything?

I think that it might make more sense to just check the lastChild, and if it exists set `lastChild.isVisible = false`

andrey-git replied:
Before my change closing more-info would just hide the element. At least in chrome camera stream  would still continue to run.

I tried to remove (without resetting the URL) the more-info-camera from the DOM, but the stream would still run.

I refactored this to use different functions for the null and non-null cases.

Regarding speed:
1) opening more-info-light [no change]
2) closing it. [sets isVisible to false, light don't care but a small amount of work is done]
3) opening it again [exactly same effort. The dom is still there and received a stateObj like before my change]

michaelarnauts replied:
This actually had cost me money! I wasn't aware that the camera stream kept running when the dialog was closed, and it kept on streaming on my mobile phone.

balloob commented on src/util/hass-util.html:
```diff
@@ -47,16 +47,20 @@
   var customEl;
 
-  if (rootEl.lastChild && rootEl.lastChild.tagName === newElementTag) {
+  if (rootEl.lastChild && (rootEl.lastChild.tagName === newElementTag || !newElementTag)) {
```
Not sure if I see what is happening here? We should only assign customEl to lastChild if it is the same tag.

andrey-git replied:
Refactored.

balloob commented on src/util/hass-util.html:
```diff
@@ -47,16 +47,20 @@
     if (rootEl.lastChild) {
       rootEl.removeChild(rootEl.lastChild);
     }
+    if (!newElementTag) return;
```
Why return here? If we just want to remove the lastChild if this is null, we should move that to the top of the function. Now the logic gets hard to follow.

andrey-git replied:
Refactored

andrey-git committed:
Split dynamicContentUpdater

Split dynamicContentUpdater into two functions intended for 'normal' flow and 'hide' flow.

andrey-git committed:
Use hassUtil.resetDynamicContent

andrey-git committed:
Fix lint

andrey-git committed:
Fix lint

balloob commented on src/util/hass-util.html:
```diff
@@ -63,12 +65,23 @@
   Object.keys(attributes).forEach(function (key) {
     customEl[key] = attributes[key];
   });
+  customEl.isVisible = true;
```
isVisible is not part of dynamicContentUpdater. dynamicContentUpdater only dynamically updates content, it doesn't mean that the content will be visible.

andrey-git replied:
I'm looking for a way to cancel the effect of a previous call to resetDynamicContent.

andrey-git replied:
I could add 'isVisible':false to stateObj before calling dynamicContentUpdater in more-info-content.html

What do you think?

balloob replied:
Yes, adding it to more-info-content.html sounds like the way to go. We wouldn't need `resetDynamicContent` - `dynamicContentUpdater` could just be used?

andrey-git replied:
Moved isVisible=true setting to  more-info-content.html

andrey-git committed:
Move isVisible to more-info-content

andrey-git committed:
Move isVisible to more-info-content

andrey-git committed:
Typo

andrey-git committed:
Avoid modifying the original state

andrey-git committed:
Put isVisible in the right place

balloob commented on src/util/hass-util.html:
```diff
@@ -69,6 +71,16 @@
 };
 
+// If root has lastChild, set 'isVisible' attribute of that child to false.
+window.hassUtil.resetDynamicContent = function (root) {
```
I think that you should inline this logic into `more-info-content.html`, it's the only place that knows about `isVisible`. Alternative, we could decide to just remove the whole `more-info` tag from the DOM when there is no stateObj, so there won't even be a need for `isVisible`

andrey-git replied:
Removing the more-info tag is not good enough - the stream continues to run. The src must be changed to non-stream to stop it.

I'll inline resetDynamicContent into more-info-content.html

balloob replied:
When we remove the child from the DOM, the image is no longer attached to the DOM either and thus it should stop streaming? Otherwise that might be a browser bug.

andrey-git replied:
I tried it with chrome desktop and mobile.

Probably the object is not destroyed when removed. Maybe there is a reference to it somewhere in our code or in Polymer.

andrey-git committed:
Inline resetDynamicContent

andrey-git committed:
Inline resetDynamicContent

andrey-git committed:
Fix lint
